src/browser/XmlHttpRequestConfigFetcher.ts

```diff
@@ -1,55 +1,92 @@
 import type { OptionsBase } from "../ConfigCatClientOptions";
-import type { IConfigFetcher, IFetchResponse } from "../ConfigFetcher";
-import { FetchError } from "../ConfigFetcher";
+import { isCdnUrl } from "../ConfigCatClientOptions";
+import type { LoggerWrapper } from "../ConfigCatLogger";
+import type { FetchRequest, IConfigCatConfigFetcher } from "../ConfigFetcher";
+import { FetchError, FetchResponse } from "../ConfigFetcher";
 
-export class XmlHttpRequestConfigFetcher implements IConfigFetcher {
-  private handleStateChange(httpRequest: XMLHttpRequest, resolve: (value: IFetchResponse) => void, reject: (reason?: any) => void) {
+export class XmlHttpRequestConfigFetcher implements IConfigCatConfigFetcher {
+  private static getFactory(): (options: OptionsBase) => IConfigCatConfigFetcher {
+    return options => {
+      const configFetcher = new XmlHttpRequestConfigFetcher();
+      configFetcher.logger = options.logger;
+      return configFetcher;
+    };
+  }
+
+  private logger?: LoggerWrapper;
+
+  private handleStateChange(httpRequest: XMLHttpRequest, isCustomUrl: boolean, resolve: (value: FetchResponse) => void, reject: (reason?: any) => void) {
     try {
       if (httpRequest.readyState === 4) {
         const { status: statusCode, statusText: reasonPhrase } = httpRequest;
 
-        if (statusCode === 200) {
-          const eTag: string | undefined = httpRequest.getResponseHeader("ETag") ?? void 0;
-          resolve({ statusCode, reasonPhrase, eTag, body: httpRequest.responseText });
-        } else if (statusCode) {
-          // The readystatechange event is emitted even in the case of abort or error.
-          // We can detect this by checking for zero status code (see https://stackoverflow.com/a/19247992/8656352).
-          resolve({ statusCode, reasonPhrase });
+        // The readystatechange event is emitted even in the case of abort or error.
+        // We can detect this by checking for zero status code (see https://stackoverflow.com/a/19247992/8656352).
+        if (statusCode) {
+          const headers = isCustomUrl ? this.getResponseHeaders(httpRequest) : getResponseHeadersDefault(httpRequest);
```

Comment:
What is the reason behind this? As I see both getResponseHeaders and getResponseHeadersDefault works the same, and I don't see any reason to treat them differently based on isCustomUrl.

Reply:
The difference is that subclassing the fetcher class and overriding `getResponseHeaders` will have any effect only in the case of a custom URL. In other words, this prevents inheritors from messing with response headers when fetching from the ConfigCat CDN.

This is the same pattern that we use in the case of `getRequestHeaders`. I'd keep it for consistency and symmetry.